### ISSUE TITLE
Is there a way to set seed while generating shap values for reproducibility?

### ISSUE THREAD

**Issue by jennykathambi90** [shap/shap#2095]
Dear @slundberg,

Thank you for this beautiful package. I have used it often but currently I have realized that whenever I generate shap values, the order of feature importance changes. I am using very few observations for modeling hence this could explain why there is so much variability.

**Comment by nklemashev** [comment]
I have encountered this problem with permutation explainer. I have written a simple code, which calculates SHAP values 100 times and checks the range for one of them. The range happens to be comparable with the mean SHAP value.

```
from sklearn.neural_network import MLPClassifier
import numpy as np
import shap

np.random.seed(42)
X = np.random.random((100, 4))
y = np.random.randint(size = (100, ), low = 0, high = 1)

model = MLPClassifier().fit(X, y)

explainer = shap.Explainer(
    model = model.predict_proba,
    masker = shap.maskers.Independent(
        data = X,
        max_samples = 100
    ),
    algorithm = 'permutation',
    feature_names = ['x_%d' % a for a in range(4)]
)

n_iter = 100
vals = np.empty((n_iter,))
for i in range(n_iter):
    vals[i] = explainer(
        X[:1, :],
        max_evals = 100,
        main_effects = False,
        error_bounds = False,
        batch_size = 'auto',
        outputs = None,
        silent = True
    ).values[0, 0, 0]

print(vals.max() - vals.min())
print(vals.mean())
```

This prints (for one of my runs):
0.0005101038560628107
-0.0006528665799572651

The simplest solution would be just to fix the seed in shap\explainers\_permutation.py by adding `np.random.seed(42)` in the method `explain_row`. After this addition the code above shows zero variance in SHAP values. Adding seed as a parameter will require a bit more changes. There is currently (I have tested with SHAP 0.39.0) no way to fix seed for permutation algorithm.

It would be great, if shap allowed setting up seed or, at least, fixed it to some value for having reproducibility. It is hard to trust explanations, that are random. @slundberg, could you, please, consider this?

**Comment by GiovannaNicora** [comment]
Having the same issue, solved as @nklemashev suggests. @slundberg it would be great to be able to set the seed for reproducibility

**Comment by CarlaFernandez** [comment]
Rooting for this one too.
EDIT: Apparently, there's already an option to set the seed, at least in the KernelExplainer constructor. Cool :)

**Comment by M-Schreurs** [comment]
The same edit works for the tree explainer by adding `np.random.seed(42)` to the shap_values function in the _tree.py file.

**Comment by marcelobeckmann** [comment]
Unfortunately that is not working for me, I set the random seed, the input values and model are the same, but I keep getting disparate features and values, even with np.random.seed(42).

I see this as a serious problem, because for the same data and predictive model, we'll land in different conclusions. 

We should not even need to set a random seed, and still, get convergent results in terms of feature importance on every run. I mean "convergent results" as: the top most important features from another run should be there between the top n features of another run, but this is not happening.

Hopefully someone could shed some light on this behavior.

**Comment by marcelobeckmann** [comment]
I managed to guarantee a stable set of Shapley values, by increasing the parameter max_evals in the PermutationExplainer. Initially I was using max_evals=1400 (the minimum recommended for 700 features), and there was no convergence (i.e., every time I've got a completely different set of top ranked features, even setting the random seed). I pushed that to max_evals=30_000, and the values are converging and consistent now, and I'm not setting the random seed.

With his I understand **there is no issue here**, as long you increase the number of max_evals (if you want something good and consistent, you have to pay the price, in terms of CPU time). 

To set a random seed (rather than increase number of evaluations) is not beneficial in this context, as it won't give you any guarantee that those Shapley values are the best set of top features, in relation to that Machine Learning model you are trying to explain. Instead, it's better to increase the max_evals and get a convergent and robust set of Shapley values.

I think this ticket can be closed, as there is no reason to set a random seed here, instead, we have to secure convergent top ranked features, but don't expect this will come for free.

**Comment by CangyuanLi** [comment]
I am not sure what exact version it was added, but in the version I am using, shap==0.45.0, shap.Explainer has a seed parameter you can set, so it seems like this functionality has been added.

**Comment by ZakaryaROUZKI** [comment]
@CangyuanLi  can you share whats the parameter name, and where you did find it ?

**Comment by rouzki** [comment]
@CangyuanLi can you share whats the parameter name, and where you did find it ?

**Comment by CangyuanLi** [comment]
> @CangyuanLi can you share whats the parameter name, and where you did find it ?

https://shap.readthedocs.io/en/latest/generated/shap.Explainer.html

You can just use shap.Explainer(seed=1)
